### PR TITLE
Re-design proteoma

### PR DESCRIPTION
**Pull request by p3rcypj**
### :pushpin: References

-   **Issue:** Closes https://app.clickup.com/t/2d9gwjh

### :memo: Implementation

Modified design of the proteome. The proteome was removed from the main app and is now in the covid table. The proteome is now rendered using SVG.
- The proteome is hidden by default. It will appear when you click the "View Proteome" button in the covid table header.
- Clicking on a gene or protein will trigger the search with its term and will also highlight the search box.
- Clicking will also send an event in GA.
- When the search box gets focus, it will remove the yellow highlight from the searchbox.
- Added styles on hover and on active.
- Added a small tag that shows if the search is running. For now, it is implemented with a timeout (700ms) so the user can read it.
- Hover on a gene or protein will change the title of the proteoma with is term; hover on gen will also show its proteins.

### :art: Screenshots
image
image
image

### CODE REVIEW

File: app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+
+    return (
+        <path
+            className={classStyle}
+            d={def}
+            onMouseEnter={() => setTitle(<span>{name}</span>)}
+            onClick={() => {
+                setLoading(true);
+                setSearch(name);
+                setLoading(false);
```
Comment (p3rcypj) [inline app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx:366]: @tokland How can I retrieve when the DataGrid has stopped looking for something?

File: app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+import styled from "styled-components";
+import _ from "lodash";
+import i18n from "../../utils/i18n";
+import { sendAnalytics } from "../../utils/analytics";
+
+interface SVGProteomaProps {
+    search: string;
+    setSearch: (value: string) => void;
+    visible: { orf1a?: boolean; orf1b?: boolean };
+    setVisible: (visible: { orf1a?: boolean; orf1b?: boolean }) => void;
```
Comment (tokland) [inline app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx:11]: The type `{ orf1a?: boolean; orf1b?: boolean }` is repeated 3 times in the app, let's create a type alias.

File: app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+
+    return (
+        <Container>
+            {/*Using relative -> absolute for having svg above title. So the title can be easily put in center*/}
+            <Layer>{title}</Layer>
+            <Layer>
+                <SVG xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1000 1000">
+                    {/*Wraps all the graphics*/}
+                    <g
+                        onMouseLeave={() => {
```
Comment (tokland) [inline app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx:30]: `useCallback` (here and other places)

File: app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+                                setVisible({ orf1a: true, orf1b: false });
+                            }}
+                        >
+                            <ProteomaPath
+                                stateSetters={stateSetters}
+                                name="ORF1a"
+                                classStyle="orf1a"
+                                def="M754,500a256.17,256.17,0,0,1-2.84,39.79A243.33,243.33,0,0,1,739,585.89l-45-16.4a204.32,204.32,0,0,0,9.52-37.28A208.29,208.29,0,0,0,706,500c0-2.5,0-4.92-.13-7.21A205.82,205.82,0,0,0,501,294V246a255.53,255.53,0,0,1,47.43,4.61,252.48,252.48,0,0,1,89.88,36.32A253.31,253.31,0,0,1,753.84,491.11C754,493.94,754,496.92,754,500Z"
+                            />
+                            {visible.orf1a && (
```
Comment (tokland) [inline app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx:54]: The component is huuuge. Let's split it a little bit. It's usually ok, as we discussed, to add internally used components in the same file, but when a component grows that much, I think it's more clear to create another file. You may create a folder to contain them all: `components/proteome/SVGProteome.tsx` + all inner components.

File: app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx [hunk 5]
```diff
@@ -0,0 +1,498 @@
...
+                setProteomaSelected(true);
+                sendAnalytics({
+                    type: "event",
+                    category: "proteoma",
+                    action: "select",
+                    label: name,
+                });
+                setTimeout(() => {
+                    setLoading(false);
+                }, 700);
```
Comment (tokland) [inline app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx:375]: Nice!

**Review by tokland — COMMENTED** [review]
Cool! Feature worked nicely as far as I can see. 

I won't say much on the design front, just that there is maybe too much top/bottom margins?

Some minor notes about the code:

**Comment by p3rcypj** [comment]
> I won't say much about the design, just: there is maybe too much top/bottom margins?

Yeah, we all agree. Those margins are produced by the n+1 rings. The SVG 1000px square size, and the inner ring actually is only 510px diameter... so you can imagine where the other 490px goes.

There are some workarounds for those margins but will break the user experience. For example we could increase the svg size only when gen is hovered, but the table will get lowered on screen without transition. We could also only trim the svg size to the real size of the graphics ~850px (because orf1b doesn't have that many rings). But this will cause to move the circle and be positioned a bit above from the real center.

A complex solution could be to only show the inner ring and when gen gets hover, instead of increase size of svg, reduce the size of the inner ring with an animation and then show the proteins. Like a zoom out.

@adrianq

File: app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx [hunk 6]
```diff
@@ -0,0 +1,498 @@
+import React from "react";
+import styled from "styled-components";
+import _ from "lodash";
+import i18n from "../../utils/i18n";
+import { sendAnalytics } from "../../utils/analytics";
+
+interface SVGProteomaProps {
```
[the remark this answers is not in the document]
Comment (p3rcypj) [reply app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx:7]: Yeah, sorry my bad.

File: app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+    setVisible: (visible: { orf1a?: boolean; orf1b?: boolean }) => void;
+    title: React.ReactNode;
+    setTitle: (value: React.ReactNode) => void;
+    setProteomaSelected: (value: boolean) => void;
+}
+
+export const SVGProteoma: React.FC<SVGProteomaProps> = React.memo(props => {
+    const { search, setSearch, visible, setVisible, title, setTitle, setProteomaSelected } = props;
+    const [loading, setLoading] = React.useState(false);
+    const stateSetters = { setSearch, setTitle, setProteomaSelected, setLoading };
```
[the remark this answers is not in the document]
Comment (p3rcypj) [reply app/assets/javascripts/covid19/src/webapp/components/SVGProteoma.tsx:20]: Thanks for the advice!

**Comment by adrianq** [comment]
Yeah, no worries too much about the margins for now. Let's see what other information the client want to show and we can then refine the styles a bit.
The PR is working fine. Good job! Just one minor thing, I would keep the message on the bottom right corner a bit more (1.5 secs?) and I would make it a bit bigger and bold.
I will re-test the PR once you push the changes required by @tokland and merge if everything is still working fine

File: app/assets/javascripts/covid19/src/webapp/components/proteome/SVGProteome.tsx [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+                {i18n.t("SARS-CoV-2")}
+                <br />
+                {i18n.t("Proteome")}
+            </span>
+        );
+        restoreVisible();
+    }, [setTitle, restoreVisible]);
+    const onClickOrf1a = React.useCallback(() => {
+        setVisible({ orf1a: true, orf1b: false });
+    }, [setVisible]);
```
Comment (tokland) [inline app/assets/javascripts/covid19/src/webapp/components/proteome/SVGProteome.tsx:49]: [best practices] [no need to change] the name of callback should reflect *what* they do, not the event that fired. Typically that's more declarative:

```
const setOrf1aVisible = useCallback(...)
...
return <Link onClick={setOrf1aVisible} />```
```

File: app/assets/javascripts/covid19/src/webapp/components/structures-table/SearchBar.tsx [hunk 9]
```diff
@@ -102,6 +119,7 @@ export const SearchBar: React.FC<SearchBarProps> = React.memo(props => {
                             {...params}
                             variant="outlined"
                             value={stateValue}
+                            onFocus={() => setProteomeSelected(false)}
```
Comment (tokland) [inline app/assets/javascripts/covid19/src/webapp/components/structures-table/SearchBar.tsx:122]: `useCallback`

**Review by tokland — COMMENTED** [review]
I like the idea of abstracting the paths as data structures! Some very minor comments:

File: app/assets/javascripts/covid19/src/webapp/components/proteome/PathGroups.tsx [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        name: "NSP2",
+        def:
+            "M607.34,231.59l-8.62,21.34a263.62,263.62,0,0,0-47-13.9l4.39-22.59A288.53,288.53,0,0,1,607.34,231.59Z",
+    },
+    {
+        name: "NSP1",
+        def:
+            "M554.15,216.06l-4.39,22.59A268,268,0,0,0,501,234V211A291.41,291.41,0,0,1,554.15,216.06Z",
+    },
+];
```
[the remark this answers is not in the document]
Comment (p3rcypj) [reply app/assets/javascripts/covid19/src/webapp/components/proteome/PathGroups.tsx:92]: Thanks!

File: app/assets/javascripts/covid19/src/webapp/components/app/Root.tsx [hunk 11]
```diff
@@ -1,21 +1,67 @@
...
                 <Wrapper>
-                    <h1>
-                        <b>{i18n.t("SARS-CoV-2 in structure databases")}</b>
-                    </h1>
+                    <h1>{i18n.t("SARS-CoV-2 in structure databases")}</h1>
+                    <Button
+                        variant="outlined"
+                        color="inherit"
+                        startIcon={isAccordionExpanded ? <VisibilityOffIcon /> : <VisibilityIcon />}
+                        onClick={() => {
```
[the remark this answers is not in the document]
Comment (p3rcypj) [reply app/assets/javascripts/covid19/src/webapp/components/app/Root.tsx:29]: I'm sorry I forgot it in the Root 😓

File: app/assets/javascripts/covid19/src/webapp/components/app/Root.tsx [hunk 12]
```diff
@@ -1,21 +1,67 @@
...
+} from "@material-ui/icons";
 import { StructuresTable } from "../structures-table/StructuresTable";
+import { SVGProteome, VisibleGen } from "../proteome/SVGProteome";
+import { useBooleanState } from "../../hooks/useBoolean";
+import i18n from "../../../utils/i18n";
 
 export const Root: React.FC = React.memo(() => {
+    const [search, setSearch] = React.useState("");
+    const [isProteomeSelected, setProteomeSelected] = React.useState(false);
+    const [title, setTitle] = React.useState<React.ReactNode>(<></>);
```
[the remark this answers is not in the document]
Comment (p3rcypj) [reply app/assets/javascripts/covid19/src/webapp/components/app/Root.tsx:16]: I needed for passing a <br/> for the default string. What would be a better workaround?

[the remark this answers is not in the document]
Comment (tokland) [reply app/assets/javascripts/covid19/src/webapp/components/app/Root.tsx:16]: It seems that the title has two states: "SARS-Proteome" and "${name}", right?. What about? `type Title = {type: "default"} | {type: "name", name: string}`. Union types are also very useful for view states.

**Comment by p3rcypj** [comment]
Thank you @tokland for the review. I'm sorry for those useCallback misses. I'll improve the next time.

**Comment by tokland** [comment]
No problem, I also forget some times about them, the temptation is always there :)

File: app/assets/javascripts/covid19/src/webapp/components/app/Root.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (p3rcypj) [reply app/assets/javascripts/covid19/src/webapp/components/app/Root.tsx:16]: Ohh I see! Thank you!

**Review by tokland — APPROVED** [review]
All good, thx @p3rcypj !

**Pull request merged**